app/src/ui/diff/text-diff.tsx

```diff
@@ -480,17 +499,138 @@ export class TextDiff extends React.Component<ITextDiffProps, {}> {
       }
     }
 
-    const items = [
+    const items: IMenuItem[] = [
       {
         label: 'Copy',
         action,
         enabled: this.onCopy && isTextSelected,
       },
     ]
 
+    const discardMenuItems = this.buildDiscardMenuItems(instance, event)
+    if (discardMenuItems !== null) {
+      items.push({ type: 'separator' }, ...discardMenuItems)
+    }
+
     showContextualMenu(items)
   }
 
+  private buildDiscardMenuItems(
+    editor: CodeMirror.Editor,
+    event: Event
+  ): ReadonlyArray<IMenuItem> | null {
+    if (!enableDiscardLines()) {
+      return null
+    }
+
+    const file = this.props.file
+
+    if (this.props.readOnly || !canSelect(file)) {
+      // Changes can't be discarded in readOnly mode.
+      return null
+    }
+
+    if (!(event instanceof MouseEvent)) {
+      // We can only infer which line was clicked when the context menu is opened
+      // via a mouse event.
+      return null
+    }
+
+    if (!this.props.onDiscardChanges) {
+      return null
+    }
+
+    const lineNumber = editor.lineAtHeight(event.y)
+    const diffLine = diffLineForIndex(this.props.diff.hunks, lineNumber)
+    if (diffLine === null || !diffLine.isIncludeableLine()) {
+      // Do not show the discard options for lines that are not additions/deletions.
+      return null
+    }
+
+    const range = findInteractiveDiffRange(this.props.diff.hunks, lineNumber)
+    if (range === null) {
+      return null
+    }
+    const hunkType = this.getHunkType(range)
+    if (hunkType === null) {
+      return null
+    }
+
+    // When user opens the context menu from the hunk handle, we should
+    // discard the range of changes that from that hunk.
+    if (targetHasClass(event.target, 'hunk-handle')) {
+      return [
+        {
+          label: this.getDiscardLabel(hunkType, range.to - range.from + 1),
+          action: () => this.onDiscardChanges(file, range.from, range.to),
+        },
+      ]
+    }
+
+    // When user opens the context menu from a specific line, we should
+    // discard only that line.
+    if (targetHasClass(event.target, 'diff-line-number')) {
+      // We don't allow discarding individual lines on hunks that have both
+      // added and modified lines, since that can lead to unexpected results
+      // (e.g discarding the added line on a hunk that is a 1-line modification
+      // will leave the line deleted).
+      return [
+        {
+          label: this.getDiscardLabel(hunkType, 1),
+          action: () => this.onDiscardChanges(file, lineNumber),
+          enabled: hunkType !== HunkType.Modified,
+        },
+      ]
+    }
+
+    return null
+  }
+
+  private getHunkType(range: { from: number; to: number }): HunkType | null {
```

Comment:
Makes sense! I think the naming that you proposed is much better than the current one as well 😅 

re. checking each line, I tried to avoid having to do so because new added files (or removed ones) there's only going to be a single hunk (or range composition) and we'll need to iterate through all file lines. Anyways since the check is super simple and it's only executed when right clicking I don't think it should have any noticeable impact on performance (I'm going to do a quick check with a file with a few thousand lines just to be sure).

Reply:
I've done some tests with out current `yarn.lock` file which is 11k lines (I think this could be a relatively common scenario, where a user wants to discard a `yarn.lock` file), and while the results are worse than I expected they're still not terribly bad.

It took between 2ms in the best case (probably with all codecaches warned in v8) and 20ms in the worse case (these results are on an average Intel i5 Windows laptop).

Something a bit worrying is that we have to call `diffLineForIndex()` for each line in the range, and this method iterates through all hunks on the file, which can potentially lead to an `O(n * m)` operation (it's a quite rare scenario, since this would require selecting a hunk with a ton of lines inside a file with a ton of hunks, but still something to be aware).

Main problem here is `getHunkType()` accepts a range, which can potentially contain lines from multiple hunks, although we know for sure (if I'm not wrong) that in the discard scenario you can only discard a hunk at a time. I'm going to try to change slightly the API of this method to be more restrictive and avoid this `O(n * m)` operation

Reply:
I've ended up addressing this in 754f85b49: by moving this calculation to the `findInteractiveDiffRange()` method we make it costless (at least in big o terms).

This could add (an unapreciable tbh) overhead to other methods using `findInteractiveDiffRange()`, but since this is only used for calculating selections on partial commits it should be fine.